pkg/controller/disruption/disruption.go

```diff
@@ -770,16 +766,15 @@ func (dc *DisruptionController) syncStalePodDisruption(ctx context.Context, key
 		return nil
 	}
 
-	podApply := corev1apply.Pod(pod.Name, pod.Namespace).
-		WithStatus(corev1apply.PodStatus()).
-		WithResourceVersion(pod.ResourceVersion)
-	podApply.Status.WithConditions(corev1apply.PodCondition().
-		WithType(v1.DisruptionTarget).
-		WithStatus(v1.ConditionFalse).
-		WithLastTransitionTime(metav1.Now()),
-	)
-
-	if _, err := dc.kubeClient.CoreV1().Pods(pod.Namespace).ApplyStatus(ctx, podApply, metav1.ApplyOptions{FieldManager: fieldManager, Force: true}); err != nil {
+	newPod := pod.DeepCopy()
+	updated := apipod.UpdatePodCondition(&newPod.Status, &v1.PodCondition{
+		Type:   v1.DisruptionTarget,
+		Status: v1.ConditionFalse,
+	})
+	if !updated {
+		return nil
+	}
+	if _, err := dc.kubeClient.CoreV1().Pods(pod.Namespace).UpdateStatus(ctx, newPod, metav1.UpdateOptions{}); err != nil {
```

Comment:
why isn't this using SMP patch? doing an update with no resource version set in UpdateOptions can stomp other pod status changes made in the meantime

Reply:
The intention of using update here is to fail on conflicts. I describe the possible scenario here: https://github.com/kubernetes/kubernetes/pull/121103#pullrequestreview-1685540049.

Basically, if we don't check resource  version here, it is possible (but unlikely) the pod will be deleted when `DisruptionTarget=False`, if the disruption controller switches the status from True to False due to categorizing it as stale after 2min, and in this time another controller updates the condition (with new LastTransitionTime).

Reply:
I double checked that the concurrent update of pod performed analogously as  in disraption.go fails. I used the following test:
```golang
func TestUpdateAfterPatch(t *testing.T) {
	testCtx := setup(t, "concurrent-updates")
	cs := testCtx.ClientSet
	pod := &v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name: "testpod",
		},
		Spec: v1.PodSpec{
			Containers: []v1.Container{
				{
					Name:  "my",
					Image: "my",
				},
			},
		},
	}
	pod, err := cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
	if err != nil {
		t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
	}
	newStatus := pod.Status.DeepCopy()
	newStatus.Phase = v1.PodFailed
	if err := util.PatchPodStatus(testCtx.Ctx, cs, pod, newStatus); err != nil {
		t.Fatalf("Error %v, while patching pod (patch1): %v", err, klog.KObj(pod))
	}
	podNew2 := pod.DeepCopy()
	apipod.UpdatePodCondition(&podNew2.Status, &v1.PodCondition{
		Type:   v1.DisruptionTarget,
		Status: v1.ConditionFalse,
	})
	if _, err := cs.CoreV1().Pods(pod.Namespace).UpdateStatus(testCtx.Ctx, podNew2, metav1.UpdateOptions{}); err != nil {
		t.Fatalf("Error %v, while updating pod (update2): %v", err, klog.KObj(pod))
	}
}
```

This fails with ``Error Operation cannot be fulfilled on pods "testpod": the object has been modified; please apply your changes to the latest version and try again, while updating pod (update2): concurrent-updates02bbb0a7-c74f-42ef-a662-373f57ee19b0/testpod`.

So I think the current behavior reflects the sentiment of staying on the safe side expressed also in this [comment](https://github.com/kubernetes/kubernetes/pull/121103#issuecomment-1768951983).

Reply:
I think that the updates performed by disraption.go should be exception, as they require 2min of the DisruptionTarget being stale. Because it is rare, and negative path (the other controlled to succeed with Patch, but fail with Delete), I think correctness might be more important than performance. Still, I'm ok with both approaches, just outlining the reasoning.

Reply:
> Doing an update like this with no resourceVersion set in UpdateOptions

you're correct, the resourceVersion is in the pod being sent to the update, not in the UpdateOptions

having this controller fail safe and requeue in a conflict case sgtm